circuit/src/gadgets/range.rs

```diff
@@ -0,0 +1,471 @@
+
+				// Assign cumulative sum such that
+				//          z_i = 2^{K}⋅z_{i + 1} + a_i
+				// => z_{i + 1} = (z_i - a_i) / (2^K)
+				//
+				// For `element` = a_0 + 2^10 a_1 + ... + 2^{120} a_{12}}, initialize z_0 =
+				// `element`. If `element` fits in 130 bits, we end up with z_{13} = 0.
+				let mut z = x.clone();
+				let inv_two_pow_k = F::from(1u64 << K).invert().unwrap();
+				for word in words {
+					// Enable q_lookup on this row
+					ctx.enable(selector.clone())?;
+
+					// z_next = (z_cur - m_cur) / 2^K
+					let z_next = z.value().zip(word).map(|(z, word)| (*z - word) * inv_two_pow_k);
+
+					// Assign z_next
+					ctx.next();
+					z = ctx.assign_advice(mock_common.common.advice[0], z_next)?;
+
+					zs.push(z.clone());
+				}
+
+				ctx.constrain_to_constant(zs.last().unwrap().clone(), F::zero())?;
+
+				Ok(self.x.clone())
+			},
+		)
+	}
+}
+
+/// RangeChipsetConfig
+#[derive(Debug, Clone)]
+pub struct RangeChipsetConfig {
+	running_sum_selector: Selector,
+	short_word_selecor: Selector,
+}
+
+impl RangeChipsetConfig {
+	/// Construct a new instance
+	pub fn new(running_sum_selector: Selector, short_word_selecor: Selector) -> Self {
+		Self { running_sum_selector, short_word_selecor }
+	}
+}
+
+/// RangeChipset
+#[derive(Debug, Clone)]
+pub struct RangeChipset<F: FieldExt, const K: usize, const N: usize, const S: usize> {
+	x: AssignedCell<F, F>,
+}
+
+impl<F: FieldExt, const K: usize, const N: usize, const S: usize> RangeChipset<F, K, N, S> {
+	/// Constructs new chipset
+	pub fn new(x: AssignedCell<F, F>) -> Self {
+		Self { x }
+	}
+}
+
+impl<F: FieldExt, const K: usize, const N: usize, const S: usize> MockChipset<F>
+	for RangeChipset<F, K, N, S>
+{
+	type Config = RangeChipsetConfig;
+	type Output = AssignedCell<F, F>;
+
+	fn synthesize(
+		self, mock_common: &MockCommonConfig, config: &RangeChipsetConfig,
+		mut layouter: impl halo2::circuit::Layouter<F>,
+	) -> Result<Self::Output, halo2::plonk::Error> {
+		// First, check if x is less than 256 bits
+		let range_chip = LookupRangeCheckChip::<F, K, N>::new(self.x.clone());
+		let x = range_chip.synthesize(
+			mock_common,
+			&config.running_sum_selector,
+			layouter.namespace(|| "range check"),
+		)?;
+
+		// Take first "num_bits" bits of `element`.
+		let num_bits = K * N;
+		let bits = x.value().map(|element| {
+			fe_to_le_bits(element.clone()).into_iter().take(num_bits).collect::<Vec<_>>()
+		});
+
+		// Chunk the "bits" into K-bit words.
+		let words = bits
+			.map(|bits| {
+				bits.chunks_exact(K)
+					.map(|word| F::from(le_bits_to_u64::<K>(&(word.try_into().unwrap()))))
+					.collect::<Vec<_>>()
+			})
+			.transpose_vec(N);
+
+		// Take away the last word of `element`(x) & check if it is 4 bit
+		let last_word = words[N - 1];
+
+		let last_word_cell = layouter.assign_region(
+			|| "last word of x",
+			|region| {
+				let mut ctx = RegionCtx::new(region, 0);
+				ctx.assign_advice(mock_common.common.advice[0], last_word)
```

Comment:
Yes, I thought about this.

I updated code so that the `LookupRangeCheckChip` returns both `x` and its "last word".
Here, the "last word" is a pre-ultimate element of `running_sum` check array.
Example: 
```
	// x:   1111111100000001
        // a_0: 11111111
        // a_1: 00000001
        //
        // z_0 = x
        // z_1 = (z_0 - a_0) / (2 ^ K)
        // z_2 = (z_1 - a_1) / (2 ^ K)
        //
        // [1111111100000001, 00000001, 0]
```
(In the example, the bits are little-endian ones. K = 8)
In this example, we can see that the "last word"(`00000001`) is the second-last element of `running_sum` check array.